common/src/main/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketCoreImpl.kt

```diff
@@ -0,0 +1,535 @@
+import io.homeassistant.companion.android.common.data.websocket.impl.entities.TemplateUpdatedEvent
+import io.homeassistant.companion.android.common.data.websocket.impl.entities.TriggerEvent
+import java.io.IOException
+import java.util.Collections
+import java.util.concurrent.atomic.AtomicLong
+import kotlin.coroutines.resumeWithException
+import kotlin.time.Duration.Companion.seconds
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.CoroutineStart
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.ProducerScope
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.channels.consumeEach
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.consumeAsFlow
+import kotlinx.coroutines.flow.shareIn
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withTimeoutOrNull
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.Response
+import okhttp3.WebSocket
+import okhttp3.WebSocketListener
+import okio.ByteString
+import okio.ByteString.Companion.toByteString
+import timber.log.Timber
+
+internal class WebSocketCoreImpl(
+    private val okHttpClient: OkHttpClient,
+    private val serverManager: ServerManager,
+    private val serverId: Int,
+) : WebSocketCore, WebSocketListener() {
+
+    private val coroutineExceptionHandler = CoroutineExceptionHandler { ctx, err -> Timber.e(err, "Uncaught exception in WebSocketCoreImpl") }
+
+    private val ioScope = CoroutineScope(Dispatchers.IO + Job() + coroutineExceptionHandler)
+
+    private val activeMessages = Collections.synchronizedMap(mutableMapOf<Long, WebSocketRequest>())
+
+    // Each message that we send needs a unique ID to match it to the answer
+    private val id = AtomicLong(1)
+    private var connection: WebSocket? = null
+    private var connectionState: WebSocketState? = null
+    private var connectionHaVersion: HomeAssistantVersion? = null
+    private val connectedMutex = Mutex()
+
+    /**
+     * A [CompletableDeferred] that signals the establishment and authentication of a WebSocket connection.
+     *
+     * This deferred value completes with `true` upon successful connection and authentication.
+     * It completes with an [Exception] if [onFailure] is invoked.
+     * It completes with an [AuthorizationException] if the authentication fail.
+     *
+     * A new instance is created when the connection closed. It is not when the authentication fail because it
+     * means it cannot recover without new credentials.
```

Comment:
@jpelgrom for the authentication what will happen a new instance of the Websocket needs to be recreated?

Reply:
I don't really follow your comment here or in the code? If authentication fails the server always closes the websocket.

Reply:
Nevermind. I was wondering what is the behavior for an authentication failure. But now it's clearer for me that we are not retrying to connect on auth failure since we probably need new creds.